src/backend/src/filesystem/hl_operations/hl_mkdir.js

```
@@ -271,8 +271,32 @@ class HLMkdir extends HLFilesystemOperation {
             });
         }
 
+        // Unify the following formats:
+        // - full path: {"path":"/foo/bar", args...}, used by apitest (./tools/api-tester/apitest.js)
+        // - parent + path: {"parent": "/foo", "path":"bar", args...}, used by puter-js (puter.fs.mkdir("/foo/bar"))
+        if ( !values.parent && values.path ) {
+            values.parent = await fs.node(new NodePathSelector(_path.dirname(values.path)));
+            values.path = _path.basename(values.path);
+        }
+
         let parent_node = values.parent || await fs.node(new RootNodeSelector());
+        if ( parent_node.isRoot ) {
```

What is the reason for changing this from `top_parent` to `parent_node`? What happens differently?

The reason this was `top_parent` before is because `top_parent` was the node immediately above the new directory being created, which is the only case when we want to throw this error (i.e. you're trying to make a new directory immediately under `/`, which isn't allowed). Has this changed?

---

yeah, `top_parent` is the right choice here